### DISCUSSION TITLE
Support for grouping of tables in the Admin UI

### DISCUSSION THREAD

**Discussion by sumitsharansatsangi** [piccolo-orm/piccolo_admin#199] [Ideas]
For a project with lots of tables  under different app requires grouping of tables.

Django provide it by creating the section with the App label ( Don't know,  is there an option to customise the name too, haven't explored).

In piccolo, if grouping of table can be provided with the custom naming support (default with the App label)  can be very helpful.

You can also provide an option to choose grouping of tables or not too.

**Comment by dantownsend** [comment]
Yes, it's an interesting idea.

You can override the table name:

```python
class Post(Table, tablename="blog_post"):
    ...
```

But I think what you're suggesting is dividing the sidebar in Piccolo Admin by app would be useful.

<img width="235" alt="sidebar_apps" src="https://user-images.githubusercontent.com/350976/184320641-507a1b48-22fd-4876-87ec-74bbf67c716b.png">

We could do something like:

```python
create_admin(tables=[
    TableConfig(Director, group='Movies'),
    TableConfig(Movie, group='Movies'),
    TableConfig(Studio, group='Movies'),
    TableConfig(Band, group='Music'),
    TableConfig(Manager, group='Music'),
])
```

**Reply by sumitsharansatsangi** [reply]
Yeah, I was talking about this.
<img src="https://user-images.githubusercontent.com/350976/184320641-507a1b48-22fd-4876-87ec-74bbf67c716b.png" >

Yeah, The following will be good. 
```python
create_admin(tables=[
    TableConfig(Director, group='Movies'),
    TableConfig(Movie, group='Movies'),
    TableConfig(Studio, group='Movies'),
    TableConfig(Band, group='Music'),
    TableConfig(Manager, group='Music'),
])
```
But default should be by the app label.
Please add the support of this at your earliest convenience.

**Reply by sumitsharansatsangi** [reply]
@dantownsend , Can you do it, please !?

**Reply by sinisaos** [reply]
@sumitsharansatsangi I can do a PR based on the `TableConfig` group parameter if @dantownsend  is ok with it. Grouping `app_name` as default is problematic because Piccolo admin doesn't use `piccolo_conf.py` where we have to pull apps from using [Finder]( https://github.com/piccolo-orm/piccolo/blob/e5c32a4810badf39fc61e465747b7343309d7e12/piccolo/apps/app/commands/show_all.py#L8). If you both agree with the `TableConfig` approach, can I push the working code? I also needed to change the Playwright tests because the `page.locator` API was changed in [v1.27.0](https://github.com/microsoft/playwright/releases/tag/v1.27.0).

**Reply by dantownsend** [reply]
@sinisaos Thanks. Yeah, I think the `TableConfig` approach is the best approach. We can add an attribute like `group_name`.

**Reply by sinisaos** [reply]
@dantownsend Great. I'll do a PR so you can see what it's all about and change something if you want. I used `menu_group` but I can change it to `group_name` if you think it's better.

**Reply by dantownsend** [reply]
menu_group is also good - up to you.

**Reply by sumitsharansatsangi** [reply]
@sinisaos ,if app_name as default is problematic , you can leave that . Grouping with the Table config approach ( as @dantownsend  suggested) is fine.

**Reply by sinisaos** [reply]
@sumitsharansatsangi #278 is PR with the `TableConfig` `menu_group`  parameter and you can try it. Usage is:
```python
from piccolo_admin.endpoints import TableConfig

movie_config = TableConfig(
    Movie,
    menu_group="Movies"
)
director_config = TableConfig(
    Director,
    menu_group="Movies"
)

create_admin([director_config, movie_config])
```

**Reply by sumitsharansatsangi** [reply]
@sinisaos , Unfortunately, I am receiving this error.
```python
Traceback (most recent call last):
  File "/home/sumit/PycharmProjects/testProj/admin.py", line 47, in <module>
    create_admin(
  File "/home/sumit/.cache/pypoetry/virtualenvs/test-proj-cxiQj_a7-py3.11/lib/python3.11/site-packages/piccolo_admin/endpoints.py", line 1098, in create_admin
    return AdminRouter(
           ^^^^^^^^^^^^
  File "/home/sumit/.cache/pypoetry/virtualenvs/test-proj-cxiQj_a7-py3.11/lib/python3.11/site-packages/piccolo_admin/endpoints.py", line 448, in __init__
    with open(os.path.join(ASSET_PATH, "index.html")) as f:
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
FileNotFoundError: [Errno 2] No such file or directory: '/home/sumit/.cache/pypoetry/virtualenvs/test-proj-cxiQj_a7-py3.11/lib/python3.11/site-packages/piccolo_admin/dist/index.html'
```

**Reply by sumitsharansatsangi** [reply]
@sinisaos, Linter is also not favorable, suggesting to remove the argument `group_admin`.

**Reply by sinisaos** [reply]
@sumitsharansatsangi You have to build frontend. In a separate terminal, go to the `admin_ui` folder and run 
```shell
// go to your_path/admin_ui
npm install  //to install packages
npm run build  //to build the dist folder
npm run serve
```

**Reply by sinisaos** [reply]
@sumitsharansatsangi You need to use argument `menu_group` not `group_admin` in `create_admin`.

**Reply by sumitsharansatsangi** [reply]
```python
Traceback (most recent call last):
  File "/home/sumit/PycharmProjects/testProj/admin.py", line 47, in <module>
    create_admin(
  File "/home/sumit/.cache/pypoetry/virtualenvs/test-proj-cxiQj_a7-py3.11/lib/python3.11/site-packages/piccolo_admin/endpoints.py", line 1098, in create_admin
    return AdminRouter(
           ^^^^^^^^^^^^
  File "/home/sumit/.cache/pypoetry/virtualenvs/test-proj-cxiQj_a7-py3.11/lib/python3.11/site-packages/piccolo_admin/endpoints.py", line 448, in __init__
    with open(os.path.join(ASSET_PATH, "index.html")) as f:
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
FileNotFoundError: [Errno 2] No such file or directory: '/home/sumit/.cache/pypoetry/virtualenvs/test-proj-cxiQj_a7-py3.11/lib/python3.11/site-packages/piccolo_admin/dist/index.html'
```

I think above is probably because of latest version of [pathspec](https://pypi.org/project/pathspec/) package. I can be wrong, but I request you to try with latest `pathspec` library.

**Reply by sinisaos** [reply]
@sumitsharansatsangi Sorry but I don't understand. That error says that you don't have `dist` folder in Piccolo Admin and you must build `frontend` using `npm run build`. You need fresh virtualenv. Then download zip from [here](https://github.com/sinisaos/piccolo_admin/tree/table_groups_in_menu) . I one terminal `cd` in `piccolo_admin` and install dependencies using pip install and after that run server with `python -m piccolo_admin.example`.
I second terminal `cd` into `piccolo_admin/admin_ui` and run `npm install`, `npm run build`(to build dist folder) and then `npm run serve`. Then you can go to http://localhost:3000 and log in to admin. 
If you want to use this in your existing project, simply copy the `piccolo_admin` (folder inside piccolo_admin) and paste it into `site-packages`, change `create_admin` to use `TableConfig(..., menu_group="Something")` and you're good to go.

**Reply by sumitsharansatsangi** [reply]
@sinisaos , The admin page has no table now, have I missed something ?
![image](https://user-images.githubusercontent.com/45959281/214618920-f3b5c49f-9411-43cd-a265-e708652832c3.png)

**Reply by sinisaos** [reply]
@sumitsharansatsangi I don't know, everything is fine on my side. I always double-check the code before PR. 

![groups](https://user-images.githubusercontent.com/30960668/214628006-ed8d6981-653d-4517-9389-4a73c7f3f961.png)

Maybe you forgot to add `menu_group` to `TableConfig`, but that shouldn't be a problem since `menu_group` is optional and if you don't use it, you get default menu without groups.
